### ISSUE TITLE
httpRequest from app-service

### ISSUE THREAD

**Issue by AFumi39** [zepp-health/zml#2]
Hello, I have an app service (service with time sensor that do action every X minutes) that reads health and device data.
I want to send those information to an external webservice.
I tried and I can send data from a page without problems. Also I successfully read sensors (tried with heartRate.getLast()) from timed app-service.
What I can't do is to send those data to a remote page, because I didn't find a way to do an HTTP request from the service.
There's a way to do that?
Or there's a way to invoke a side-service from within the app-service?

I'm running the app on Amazfit Balance (ZOS 3.0).

Thank you!

**Comment by htoooth** [comment]
‌You can use BasePage in AppService.
`AppService(BasePage({}))`

@AFumi39

**Comment by AFumi39** [comment]
> ‌You can use BasePage in AppService. `AppService(BasePage({}))`
> 
> @AFumi39

Unfortunately it doesn't works, I already tried but I receive an error (11/27 [0]16:02:07.309 I/quickjs-libc SBEngine: Mini Program Error TypeError: not a function ) when I try to launch the service.

**Comment by htoooth** [comment]
@AFumi39  Can you give me the mini-program where you encountered the issue?

**Comment by AFumi39** [comment]
> @AFumi39 Can you give me the mini-program where you encountered the issue?

[hello-world-test.zip](https://github.com/zepp-health/zml/files/13491262/hello-world-test.zip)
This is a very simple test mini-program. Can you tell me where I'm wrong?
Thanks

**Comment by htoooth** [comment]
@AFumi39  you need a sideService like this https://github.com/zepp-health/zml?tab=readme-ov-file#use-in-sideservice-module

**Comment by AFumi39** [comment]
> @AFumi39 you need a sideService like this https://github.com/zepp-health/zml?tab=readme-ov-file#use-in-sideservice-module

Sorry, I stille don't understand. This is new version with side-service. I'm trying to invoke it from my timeService, but still not working (same error). Cna you please help me to understand in which part I'm wrong?
Maybe I've to put the time-based part on side-service? How?

Thank you
[hello-world-test.zip](https://github.com/zepp-health/zml/files/13512711/hello-world-test.zip)

**Comment by AFumi39** [comment]
@htoooth any hint for me?
I'm sorry, but I don't fully understand yet the architecture of ZOS (3.0 in particular) and so I can't figure how to develop a simple app that periodically read fitness data and send it to an external webservice.
I tried many suggested methods, but yet none really works.

Thank you

**Comment by DotIN13** [comment]
I don't believe you can do this with the current API. The background app service seems to have no support for Internet connections or storage API, so my best guess is that there is no way to extract data using a background service.

You may be able to do this via a watchface though, since watchface is almost always running in the foreground, but I am not sure if you would be able to register a side service for a watchface.

**Comment by DotIN13** [comment]
Sorry for my oversight, background app service might actually be able to communicate with phone side service, as in src/core/device/app-service/bgService.js. Yet  to my knowledge there is no readily available documentation for this feature so you may have to experiment this on your own.

As for if you can periodically retrieve data from the watch solely using side service, I do not think that is possible. It either has to get the data from background app service or the device app.

**Comment by DotIN13** [comment]
I checked out your code and you seemed to be calling `this.request` directly from the AppService, and I guess this is where the `not a function` error came from. To pinpoint the error, you may start an empty AppService with BasePage and see if you still get the error to make sure the error came from the code trying to contact the phone from the AppService, and not from your code to initialize the AppService.

**Comment by DotIN13** [comment]
Your current design involving the app service is the right way to go! Now you just have to make sure the not a function error comes from the request function call, if it does, then maybe @htoooth can help us with this.

The simplest way to check is to put a logger.debug call right before the request call, and see if we can get there.

**Comment by AFumi39** [comment]
With the "last" version (that I attached), if I remove the this.request from app-service, seems that no error occurs. If I put the this.request in the Page, it works (do the http call).
I tried many way to workaround the problem, even using other project as reference (for example https://github.com/bigdigital/zeppos_watchdrip_app/tree/master_v3 ), but with no luck, ad the moment.

I'll wait for @htoooth or official devs (they said that they'll provide an easy way to do so, but there's no ETA at the moment) to find a way to resolve my problem.

Thank you

**Comment by DotIN13** [comment]
No problem, hope the devs will make it happen in no time. Oh and did you try this.call instead of request?

**Comment by AFumi39** [comment]
I don't remember if I tried this.call whit the current code. I'll try now.
Thanks!

**Comment by AFumi39** [comment]
Also with this.call I got the "not a function" error

**Comment by htoooth** [comment]
‌‌‌I'm going to deal with the memory leak issue. Do you have any questions or do you want to use the `this.request` method independently?

@AFumi39

**Comment by AFumi39** [comment]
Thank you, I just want to know if you have suggestions (or if you can give me an example) to do an httpRequest from app-service.
Unfortunately no method that I tried worked so far.
I can do it easily from a Page, but, for example, ZML library give error if I call the this.request from app-service.

Thank you

**Comment by htoooth** [comment]
In the sideService, if you want to access http, you should use `this.fetch`.   @AFumi39

**Comment by DotIN13** [comment]
@htoooth Hey htooooth, I think there might be a misunderstanding here! I think @AFumi39 was trying to run this.request in App Service, not Side Service.

**Comment by htoooth** [comment]
@DotIN13 @AFumi39 

OK. You can use `AppService(BasePage({}))` like this.

This is feasible.

**Comment by AFumi39** [comment]
@htoooth thank you, I tried `AppService(BasePage({}))`, but I get `TypeError: not a function` when I call the `this.request` in the AppService :(

**Comment by htoooth** [comment]
OK, I will check it  again @AFumi39

**Comment by htoooth** [comment]
@DotIN13 
It could be that there is a problem with your code.
![image](https://github.com/zepp-health/zml/assets/1717023/6f7976a3-2cef-4ef0-950f-8ee0987c7b51)
Here, it needs to be changed to `this.getYourData()`

**Comment by DotIN13** [comment]
Oh my, I didn't catch that! Sorry @AFumi39 ~

**Comment by AFumi39** [comment]
@htoooth @DotIN13 
oh, shit.... That was so stupid mistake that I didin't even notice until you pointed it out!!!! 😱😱🙈
I tried so many things in all my tests, that when I semplified with basic and clean code to do basic tests... I made that error without noticing...

Sorry, now it works perfectly!

Thanks ❤
